File: gameFramework/src/main/java/at/grueneis/game/framework/code/AndroidMusic.java
package at.grueneis.game.framework.code;

import java.io.IOException;

import android.content.res.AssetFileDescriptor;
import android.media.MediaPlayer;
import android.media.MediaPlayer.OnCompletionListener;
import at.grueneis.game.framework.Music;

public class AndroidMusic implements Music, OnCompletionListener
{
	MediaPlayer mediaPlayer;
	boolean isPrepared;
	
	public AndroidMusic(AssetFileDescriptor afd)
	{
		mediaPlayer = new MediaPlayer();
		try
		{
			mediaPlayer.setDataSource(afd.getFileDescriptor(), afd.getStartOffset(), afd.getLength());
			mediaPlayer.prepare();
			isPrepared = true;
			mediaPlayer.setOnCompletionListener(this);
		}
		catch (IOException exc)
		{
			throw new RuntimeException("Could not load music!");
		}
	}
	
	public void onCompletion(MediaPlayer arg0)
	{
		synchronized (this)
		{
			isPrepared = false;
		}
	}
	
	public void play()
	{
		if (mediaPlayer.isPlaying()) return;
		try
		{
			synchronized (this)
			{
				if (!isPrepared) mediaPlayer.prepare();
				mediaPlayer.start();
			}
		}
		catch (Exception exc)
		{
			exc.printStackTrace();
		}
	}
	
	public void stop()
	{
		mediaPlayer.stop();
		synchronized (this)
		{
			isPrepared = false;
		}
	}
	
	public void pause()
	{
		mediaPlayer.pause();
	}
	
	public void setLooping(boolean looping)
	{
		mediaPlayer.setLooping(looping);
	}
	
	public void setVolume(float volume)
	{
		mediaPlayer.setVolume(volume, volume);
	}
	
	public boolean isPlaying()
	{
		return mediaPlayer.isPlaying();
	}
	
	public boolean isStopped()
	{
		return !isPrepared;
	}
	
	public boolean isLooping()
	{
		return mediaPlayer.isLooping();
	}
	
	public void dispose()
	{
		if (mediaPlayer.isPlaying()) mediaPlayer.stop();
		mediaPlayer.release();
	}
	
}

File: gameFramework/src/main/java/at/grueneis/game/framework/impl/AndroidMusic.kt
package at.grueneis.game.framework.impl

import android.content.res.AssetFileDescriptor
import android.media.MediaPlayer
import android.media.MediaPlayer.OnCompletionListener
import at.grueneis.game.framework.Music
import java.io.IOException

class AndroidMusic(afd: AssetFileDescriptor) : Music, OnCompletionListener {
    var mediaPlayer: MediaPlayer = MediaPlayer()
    var isPrepared = false
    override fun onCompletion(arg0: MediaPlayer) {
        synchronized(this) { isPrepared = false }
    }

    override fun play() {
        if (mediaPlayer.isPlaying) return
        try {
            synchronized(this) {
                if (!isPrepared) mediaPlayer.prepare()
                mediaPlayer.start()
            }
        } catch (exc: Exception) {
            exc.printStackTrace()
        }
    }

    override fun stop() {
        mediaPlayer.stop()
        synchronized(this) { isPrepared = false }
    }

    override fun pause() {
        mediaPlayer.pause()
    }

    override fun setVolume(volume: Float) {
        mediaPlayer.setVolume(volume, volume)
    }

    override val isPlaying: Boolean
        get() = mediaPlayer.isPlaying
    override val isStopped: Boolean
        get() = !isPrepared
    override var isLooping: Boolean
        get() = mediaPlayer.isLooping
        set(looping) {
            mediaPlayer.isLooping = looping
        }

    override fun dispose() {
        if (mediaPlayer.isPlaying) mediaPlayer.stop()
        mediaPlayer.release()
    }

    init {
        try {
            mediaPlayer.setDataSource(afd.fileDescriptor, afd.startOffset, afd.length)
            mediaPlayer.prepare()
            isPrepared = true
            mediaPlayer.setOnCompletionListener(this)
        } catch (exc: IOException) {
            throw RuntimeException("Could not load music!")
        }
    }
}